move and enable visual tests

AngeloDanducci opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Moved and enabled the visual tests

### Details and comments

Need to verify artifact publishing occurs on failure/missing, may need to write some purposeful failures.

Fixes #9892

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=9961) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=9961) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=9961) it.</sub>

AngeloDanducci commented on test/visual/results.py:
```diff
@@ -146,14 +146,14 @@ def summary(self):
-        if len(self.mismatch) >= 2:
+        if len(self.mismatch) >= 1:
             Results._zipfiles(self.mismatch, f"{self.directory}/mismatch.zip")
             ret += (
                 f'<div><a href="{self.directory}/mismatch.zip">'
                 "Download %s mismatch results as a zip</a></div>" % len(self.mismatch)
             )
 
-        if len(self.missing) >= 2:
+        if len(self.missing) >= 1:
             Results._zipfiles(self.missing, f"{self.directory}/missing.zip")
```
This creation of zip files here never actually gets used the way it's currently set up.  The new portion added to `.azure/test_linux.yml` covers  finding and zipping.  I don't think anything else uses these summary/zipfiles methods, should they be removed?

1ucian0 replied:
yes. I think it can be removed.

1ucian0 commented on .azure/test-linux.yml:
```diff
@@ -185,10 +186,55 @@ jobs:
           env:
             SETUPTOOLS_ENABLE_FEATURES: "legacy-editable"
```
This might not be necessary

1ucian0 commented on .azure/test-linux.yml:
```diff
@@ -185,10 +186,55 @@ jobs:
-        - bash: image_tests/bin/python -m unittest discover -v test/ipynb
+        - bash: image_tests/bin/python -m unittest discover -v test/visual
           displayName: 'Run image test'
           env:
             # Needed to suppress a warning in jupyter-core 5.x by eagerly migrating to
             # a new internal interface that will be the default in jupyter-core 6.x.
             # This variable should become redundant on release of jupyter-core 6.
             JUPYTER_PLATFORM_DIRS: 1
+
```
This might not be necessary

AngeloDanducci reviewed (commented):
Is it possible that this is flakey?  Based on the run I have seen both of these as the result output.
<img width="1145" alt="Screenshot 2023-05-25 at 2 29 19 AM" src="https://github.com/Qiskit/qiskit-terra/assets/16545055/ebe8e5d5-d1a5-462f-873d-de24550a50a5">

1ucian0 commented:
This is great! thanks! 

I think it would help to find the correct result if `graph_results.tar.gz` untars in `graph_results/` instead of `results/` (same for `circuit_results.tar.gz`. Also, if a tarball is empty, probably would be less confusing if it is not included (in the current example `circuit_failures.tar.gz`)

Currently, this is the structure of artifact, once the everything is untared:
```
image_test_failure_img_diffs
├── circuitfailures
├── graphfailures
│   └── state_city.png
└── results
    ├── 16_qubit_gate_map.diff.png
    ├── 16_qubit_gate_map.png
    ├── 1_qubit_gate_map.diff.png
    ├── 1_qubit_gate_map.png
   ...
    ├── state_city.png
   ...
    ├── wide_params.png
    ├── wire_order.diff.png
    └── wire_order.png
```

AngeloDanducci commented:
Moved the visual tests to their own job and fixed the naming overlap when uncompressed.

I was looking for a control flow option to check if a folder was empty but I haven't run across anything in the azure docs that would enable that.

An alternative would be instead of running `unittest discover`, the tests could be run independently as the tars are only created on failure. In comparison currently all visual tests run at once so any failure creates the tar files (including empties).

1ucian0 commented:
> Is it possible that this is flakey? Based on the run I have seen both of these as the result output.

it looks like a rounding error. I think I fixed it in https://github.com/Qiskit/qiskit-terra/pull/9961/commits/e6528cfddb7df82b1d9f83b34536fb220d3e1e4b

jakelishman commented on azure-pipelines.yml:
```diff
@@ -143,6 +143,10 @@ stages:
           parameters:
             pythonVersion: ${{ parameters.minimumPythonVersion }}
 
+        - template: ".azure/test-visual.yml"
+          parameters:
+            pythonVersion: ${{ parameters.minimumPythonVersion }}
+
```
I think it would be better to leave this as a subcomponent of the `test-linux` run.  Our CI structure is deliberately designed such that the two-stage PR and one-stage push CI need a total of 10 VMs, which is our concurrency limit on Azure.  That lets us have 3 PRs working through their CI in parallel at peak throughput for release time - the stages are all roughly timed to take the same amount of wall time when run in parallel.

AngeloDanducci replied:
moved the visual testing back to test-linux

jakelishman commented on qiskit/visualization/state_visualization.py:
```diff
@@ -1100,6 +1100,9 @@ def generate_facecolors(x, y, z, dx, dy, dz, color):
     """
     import matplotlib.colors as mcolors
 
+    # There seems to be a rounding error in which some zero bars are negative
+    dz += 1e-10
+
```
This doesn't seem like the right way of doing things.  If the problem is that we need to clip the value to a minimum of 0, then `dz = np.clip(dz, 0, None)` is probably more like what was intended, but this doesn't look like the correct function to be doing this in - the "get the colours of the bars" part of the logic shouldn't be responsible for deciding what the heights of the bars are.  That should be wherever the bars are coming from.

AngeloDanducci replied:
used suggested clip, but in the function that calculates the height of the bars

jakelishman commented on test/visual/mpl/circuit/test_circuit_matplotlib_drawer.py:
```diff
@@ -64,19 +68,87 @@ def cwd(path):
+        black_pixels = black_and_white_version.histogram()[0]
+        return black_pixels
+
+    @staticmethod
+    def _similarity_ratio(current, expected, image_name):
+        diff_name = current.split(".")
+        diff_name.insert(-1, "diff")
+        diff_name = ".".join(diff_name)
+
+        current = Image.open(current)
+        expected = Image.open(expected)
+
+        diff = ImageChops.difference(expected, current).convert("L")
+
+        black_pixels = TestCircuitMatplotlibDrawer._get_black_pixels(diff)
+        total_pixels = diff.size[0] * diff.size[1]
+        ratio = black_pixels / total_pixels
+
+        if ratio != 1:
+            TestCircuitMatplotlibDrawer._black_or_b(diff, current, expected).save(
+                FAILURE_DIFF_DIR + FAILURE_PREFIX + image_name, "PNG"
+            )
+        else:
+            TestCircuitMatplotlibDrawer._black_or_b(diff, current, expected).save(diff_name, "PNG")
+        return ratio
```
It's surprising to me that a function called "similarity_ratio" would have the very major side effect of writing out images into hard-coded paths.  I think it would be neater to put that part of the logic into an `assertImagesEqual` method or the like - the magic-number decision to say that the cut-off for acceptable similarity is "precisely equal" is currently in two places: hidden in this function, and the `assert` that follows the call of this function.

AngeloDanducci replied:
While there is a cut off here, this does not count as a test failure, but does aggregate all non 100% matches serpately. Perhaps `NOT_IDENTICAL` would be a better name.

Changed `_similary_ratio` to `_save_diff` to more accurately reflect what the function is doing.

jakelishman commented on test/visual/mpl/circuit/test_circuit_matplotlib_drawer.py:
```diff
@@ -49,8 +50,11 @@
 
-
-RESULTDIR = os.path.dirname(os.path.abspath(__file__))
+BASEDIR = os.path.dirname(os.path.abspath(__file__))
+RESULTDIR = os.path.join(BASEDIR, "circuit_results", "")
+TEST_REFERENCE_PATH = os.path.join(BASEDIR, "references")
+FAILURE_DIFF_DIR = os.path.join(os.path.dirname(BASEDIR), "visual_test_failures", "")
+FAILURE_PREFIX = "circuit_failure_"
```
We seem to be inconsistent on whether directories are referred to as "DIR" or "PATH", and whether we use shouting snake case, or all-caps one-word.  Could we choose one for each, please?

Minor style choice, but these might be better off as `pathlib.Path` instances, then joining paths is much easier - `DIR / "my_file.png"` - to the degree that I think the `_image_path` etc methods could be skipped.  That would make the test logic a bit easier to read for me - there's lots of indirection at the moment that I have to keep in my head.

AngeloDanducci replied:
Changed consistency to shouting snake case `DIR`.

Changed to path lib.

jakelishman commented on test/visual/mpl/circuit/test_circuit_matplotlib_drawer.py:
```diff
@@ -64,19 +68,87 @@ def cwd(path):
         super().tearDown()
         mpl_close("all")
 
+    @staticmethod
+    def _image_path(image_name):
+        return os.path.join(RESULTDIR, image_name)
+
+    @staticmethod
+    def _reference_path(image_name):
+        return os.path.join(TEST_REFERENCE_PATH, image_name)
```
Minor: I'd roughly prefer these just to be inlined into the call sites with the constants becoming `pathlib.Path` instances.  It'd be the same or less typing, even at the point of call:
```
self._reference_path("my_image.png")  # current
TEST_REFERENCE_PATH / "my_image.png"  # alternate
```

AngeloDanducci replied:
Chose not to inline these for the moment - but I can see the reasoning.

jakelishman commented on test/visual/mpl/circuit/test_circuit_matplotlib_drawer.py:
```diff
@@ -64,19 +68,87 @@ def cwd(path):
+
+    @staticmethod
+    def _black_or_b(diff_image, image, reference, opacity=0.85):
+        """Copied from https://stackoverflow.com/a/30307875"""
```
I don't know if this is pre-existing, but I don't think we're satisfying the licensing conditions of StackOverflow's Creative Commons Attribution+ShareAlike if we're copying code.  If you didn't newly introduce this we don't need to fix it immediately in this PR, but we might need to look at it.

AngeloDanducci replied:
This is pre-existing, letl me know how you would like this addressed.  We can always make that a separate issue/PR.

jakelishman commented on test/visual/mpl/graph/test_graph_matplotlib_drawer.py:
Most of the comments in the `test/visual/mpl/circuit` test file apply here too, to the degree that it might be clearer to factor out much of what's a `staticmethod` here into simple stand-alone helper functions defined in `test/visual/__init__.py` that could be imported and used directly.

AngeloDanducci replied:
Refactored the graph tests to mirror the circuit test changes.

jakelishman reviewed (commented):
Thanks for looking at this. I think the Azure run actually just got lost in the webhooks aether - it happens sometimes.  The "authorisation required" thing is about a couple of the tests that run on GitHub Actions, and won't actually be affected by these changes at all.  Pushing a new commit to this branch should retrigger the webhook to get Azure going again.

It'll definitely be good to get these tests running again in CI.  I haven't been paying attention through the process of developing this PR, but just to check: have you seen any flakiness at all in any of the tests?  I don't know if there's any stochastic elements to any of the visualisations, but hopefully not.

jakelishman commented on test/visual/mpl/circuit/test_circuit_matplotlib_drawer.py:
```diff
@@ -230,7 +327,11 @@ def test_long_name(self):
+        fname = "long_name.png"
+        self.circuit_drawer(circuit, filename=fname)
+
+        ratio = self._similarity_ratio(self._image_path(fname), self._reference_path(fname), fname)
+        assert ratio == 1
```
If you were aiming to have the assertions fire only after all the tests ran, I'd suggest using `ddt` to parametrise the multi-circuit tests so there's only one assertion, or use `with self.subTest` to make internally separate test contexts.

AngeloDanducci replied:
Changed to use the unittest `assertEqual`.

Left as an assert after the fact, after changing `_similarity_ratio` to `_save_diff` I think it makes sense that the function saves the diff and returns how different it is.

AngeloDanducci commented on .azure/test-linux.yml:
```diff
@@ -11,6 +11,7 @@ parameters:
 
   - name: "testImages"
     type: boolean
+    default: true
```
moved back into test-linux

AngeloDanducci commented on .azure/test-linux.yml:
```diff
@@ -81,8 +82,6 @@ jobs:
               -e .
           displayName: "Install Terra directly"
-          env:
-            SETUPTOOLS_ENABLE_FEATURES: "legacy-editable"
```
re added for the moment

AngeloDanducci commented on .azure/test-visual.yml:
```diff
@@ -0,0 +1,91 @@
+        condition: failed()
+
+
+        
```
removed
